### ISSUE TITLE
Errors when Disconnecting?

### ISSUE THREAD

**Issue by Mheetu** [Bob74/bob74_ipl#15]
Is it normal/expected to see these errors when disconnecting from a server or have I done something wrong? 🤔

![image](https://user-images.githubusercontent.com/8495484/60665561-cad4ae00-9e32-11e9-9f6f-6cd5fbcbab39.png)

**Comment by Bob74** [comment]
No, it is not normal.
The script calls the native method `ReleaseNamedRendertarget()` but I don't know why it calls it from another file (`ReleaseNamedRendertarget.lua`) which isn't part of my scripts (maybe a FiveM client file?).
Unfortunately, I won't be able to investigate where this file is and why is it called before quite a long time (not at home atm and a bunch of IRL issues to solve first :/)

**Comment by Mheetu** [comment]
No worries, take care of yourself.

The script still appears to work, just not a fan of seeing those errors when disconnecting. I might dig into it at some point... It works and doesn't appear to be causing any issues so it's not exactly priority for me at the moment.

**Comment by Bob74** [comment]
Actually this error indicates that a GTA V variable is not released (the Rendertarget handle), so it might not erase this variable from the game's memory when you reload the script from the server or when a player disconnect.

It is not **very** important but still, it would be cleaner to properly release the memory before shutting the resource down.

**Comment by Bob74** [comment]
Can you do some research in you resources directory to see if you have the file `ReleaseNamedRendertarget.lua` somewhere?
I've started a clean server and cannot find it anywhere, even in FiveM client files.

**Comment by Mheetu** [comment]
Hey, sorry for the delay... I got super busy and didn't even notice your comment.

I did some searching and I too couldn't find that file anywhere. Likewise, my Google searches rendered no helpful results - in fact, all it found were others reporting the same error from this resource. 😕

**Comment by Mheetu** [comment]
After looking into this a bit more, my current **theory** is that it's trying to execute `ReleaseNamedRendertarget()` after parts of the game have already been unloaded when a player returns to the main menu. This would explain why it works when you stop/start the resource while in a server but not when it's stopped after leaving the server.

I would assume FiveM automatically releases those render targets, though I'm not sure... if not, this could be a potential memory leak because we're not able to execute the native function at that point in the disconnect/shutdown lifecycle and there aren't any other hooks to run code any sooner.

I'm not too familiar with mapping related aspects of the game... but out of curiousity, is there a reason why `dlc_finance/organization.lua` and `dlc_bikers/gang.lua` have an `onResourceStop` event handler and none of the others do? Similarly, why are they executing every time **any** resource is stopped and not just when `bob74_ipl` is stopped?

As a temporary workaround to merely suppress the errors, I went ahead and added the following to the top of each `onResourceStop` handler on my server... though I feel like it's a bit of a hack and not an actual solution...  

```lua
if (GetCurrentResourceName() ~= resourceName or GetCurrentServerEndpoint() == nil) then return end
```

**Comment by Bob74** [comment]
I jumped to the same conclusion last night, I think FiveM totally unloads the game when you disconnect so only functions related to CitizenFx may continue to work.

I don't remember why I let `onResourceStop` event happen for every resource though. I think it was during testing where some code was still needed on the user resource interacting with `bob74_ipl` and since I wouldn't be able to know the name of the resource for other people, I let it happen for every resource.

But it doesn't make sense anymore, everything is handled from `bob74_ipl` now.

Your workaround is actually what I should have done before releasing the update! (and I know I used this check while testing and removed it at some point).
Also, `onClientResourceStop` seems to be triggered earlier and allow the resource to stop without errors while `onResourceStop` still gives me the error.

Thanks for your help, the `ReleaseNamedRendertarget.lua` part really bugged me (and it still does) but at least we know what was causing this :)

**Comment by Mheetu** [comment]
I think the `ReleaseNamedRendertarget.lua` is just how it displays an error when a native function call fails... I had one similar with another function earlier this week, though I don't recall what it was or how to reproduce it to provide as an example. I believe it to just be an edge case or artifact with how native functions are loaded or errors are logged... something like [ [this](https://github.com/citizenfx/fivem/blob/f2a62f1c4984334b76c79c27e1cd41f2a525c8c6/data/shared/citizen/scripting/lua/natives_loader.lua#L76) ], perhaps?... though I could be entirely wrong. 🤷‍♂️

**Comment by Bob74** [comment]
I guess you're right with the way natives are loaded, kinda odd but I think they know what they're doing 😄

**Comment by Mheetu** [comment]
Interesting, I could have sworn I played with `onClientResourceStop` and it threw the same errors... It was actually one of the first things I did... 😂 Just tried it again and it seems to work...

However, I don't believe it works as you described... I think `onClientResourceStop` is not being executed at all when a player disconnects to return to the main menu. I've not looked at FiveM's source to confirm this, but if you try to `print()` something, it never appears in the client's log. Which in our particular case, is perfectly fine... provided that FiveM/the game release those resources properly.

Nonetheless, I guess we can go ahead and close this issue with switching `onResourceStop` to `onClientResourceStop` and prepending `if (GetCurrentResourceName() ~= resourceName) then return end` to it as being the super advanced and extremely complicated solution. 😂

**Comment by mikethemadkiwi** [comment]
Can confirm the error ONLY happens with the gang.lua trying to call the resourcestop.
the change to clientresourcestop is fixing it. 
seems like it's only to do with the part where it draws names on the wall.
oddly, i havent been getting an error from the executive bit calling the same code. 
but i've altered that too, to be safe!

**Comment by Bob74** [comment]
Should be fixed in the current version (2.0.8).
